docs/rfcs/700-config-features-and-defaultpairs-via-connection-string.md

```diff
@@ -0,0 +1,225 @@
+- Author: JinnyYi <github.com/JinnyYi>
+- Start Date: 2021-08-03
+- RFC PR: [beyondstorage/go-storage#700](https://github.com/beyondstorage/go-storage/issues/700)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-700: Config Features and DefaultPairs via Connection String
+
+Previous discussion:
+
+- [Connection string needs to support config StorageFeatures and DefaultPairs](https://github.com/beyondstorage/go-storage/issues/680)
+
+## Background
+
+In [GSP-90: Re-support Initialization Via Connection String](./90-re-support-initialization-via-connection-string.md), we introduced connection string to support service init and defined the format of connection string. 
+The connection string exposes basic types and not support complex types like `struct` at present. So we can't set `ServiceFeature`/`StorageFeatures` and `DefaultServicePairs`/`DefaultStoragePairs` by connection string during initialization.
+
+`Default*Pairs` is a slightly complex struct:
+
+```go
+// DefaultStoragePairs is default pairs for specific action
+type DefaultStoragePairs struct {
+	Create            []Pair
+	Delete            []Pair
+	List              []Pair
+	Metadata          []Pair
+	Read              []Pair
+	Stat              []Pair
+	Write             []Pair
+}
+```
+
+A possible format that follows the struct in connection string for default storage paris may be:
+
+ `default_storage_pairs={write:[storage_class=STANDARD_IA,key2=value],read:[]}`
+
+Obviously, It's complex for users and hard to pass in a valid string. In addition, it either exposes the internal details to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers).
+
+## Proposal
+
+I propose to support config features and default pairs in plain string format in connection string.
+
+### Features
+
+The format of features pairs in connection string is:
+
+`enable_key=value`
+
+- The whole key of the pair SHOULD have the prefix `enable_`.
+- `key` is the pair name defined in `features.toml` and supported in service, and the format SHOULD be exactly the same.
+- `value` SHOULD be `true` or `false`.
+
+So a valid connection string containing features could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&enable_loose_pair=false&enable_virtual_dir=true`
+
+The connection string above is equivalent to:
+
+```go
+store, err := s3.NewStorage(
+	ps.WithCredential("hmac:xxxx:xxxx"),
+	ps.WithName("bucket_name"),
+	ps.WithWorkDir("/prefix"),
+	s3.WithStorageFeatures(s3.StorageFeaturs{LoosePair:  true,
+		LoosePair:  true,
+		VirtualDir: true,
+    })
+)
+```
+
+### DefaultPairs
+
+The format of default pair in connection string is:
+
+`default_key=value`
+
+- The whole key of the pair SHOULD have the prefix `default_`.

```

Comment:
Maybe we don't need the prefix `default_`? Then we can reuse `pairMap` with no change and `WithStorageClass` during initialization instead of  generating `WithDefaultStrageClass`?

Reply:
As described in https://github.com/beyondstorage/go-storage/pull/700#discussion_r683939730, I don't think so.